File: Source/AlphaTab.JavaScript/Platform/JavaScript/AlphaTabWebWorker.cs
using System;
using AlphaTab.Haxe;
using AlphaTab.Haxe.Js;
using AlphaTab.Haxe.Js.Html;
using AlphaTab.Model;
using AlphaTab.Rendering;
using AlphaTab.Util;

namespace AlphaTab.Platform.JavaScript
{
    internal class AlphaTabWebWorker
    {
        private ScoreRenderer _renderer;
        private readonly DedicatedWorkerGlobalScope _main;

        public AlphaTabWebWorker(DedicatedWorkerGlobalScope main)
        {
            _main = main;
            _main.AddEventListener("message", (Action<Event>)HandleMessage, false);
        }

        public static void Init()
        {
            new AlphaTabWebWorker(Lib.Global);
        }

        private void HandleMessage(Event e)
        {
            var data = ((MessageEvent)e).Data;
            var cmd = data ? data.cmd : "";
            switch (cmd)
            {
                case "alphaTab.initialize":
                    var settings = new Settings();
                    settings.LoadFromJson(data.settings);
                    Logger.LogLevel = settings.Core.LogLevel;
                    _renderer = new ScoreRenderer(settings);
                    _renderer.PartialRenderFinished += result => _main.PostMessage(new
                    {
                        cmd = "alphaTab.partialRenderFinished",
                        result = result
                    });
                    _renderer.RenderFinished += result => _main.PostMessage(new
                    {
                        cmd = "alphaTab.renderFinished",
                        result = result
                    });
                    _renderer.PostRenderFinished += () => _main.PostMessage(new
                    {
                        cmd = "alphaTab.postRenderFinished",
                        boundsLookup = _renderer.BoundsLookup.ToJson()
                    });
                    _renderer.PreRender += resize => _main.PostMessage(new
                    {
                        cmd = "alphaTab.preRender",
                        resize = resize
                    });
                    _renderer.Error += Error;
                    break;
                case "alphaTab.invalidate":
                    _renderer.Render();
                    break;
                case "alphaTab.resizeRender":
                    _renderer.ResizeRender();
                    break;
                case "alphaTab.setWidth":
                    _renderer.Width = data.width;
                    break;
                case "alphaTab.renderScore":
                    var score = JsonConverter.JsObjectToScore(data.score, _renderer.Settings);
                    RenderMultiple(score, data.trackIndexes);
                    break;
                case "alphaTab.updateSettings":
                    UpdateSettings(data.settings);
                    break;
            }
        }

        private void UpdateSettings(object json)
        {
            var settings = new Settings();
            settings.LoadFromJson(json);
            _renderer.UpdateSettings(settings);
        }

        private void RenderMultiple(Score score, int[] trackIndexes)
        {
            try
            {
                _renderer.RenderScore(score, trackIndexes);
            }
            catch (Exception e)
            {
                Error("render", e);
            }
        }

        private void Error(string type, Exception e)
        {
            Logger.Error(type, "An unexpected error occurred in worker", e);

            var error = Json.Parse(Json.Stringify(e));

            dynamic e2 = e;

            if (e2.message)
            {
                error.message = e2.message;
            }

            if (e2.stack)
            {
                error.stack = e2.stack;
            }

            if (e2.constructor && e2.constructor.name)
            {
                error.type = e2.constructor.name;
            }

            _main.PostMessage(new
            {
                cmd = "alphaTab.error",
                error = new
                {
                    type = type,
                    detail = error
                }
            });
        }
    }
}

File: src/platform/javascript/AlphaTabWebWorker.ts
import { JsonConverter } from '@src/model/JsonConverter';
import type { Score } from '@src/model/Score';
import type { IWorkerScope } from '@src/platform/javascript/IWorkerScope';
import { type FontSizeDefinition, FontSizes } from '@src/platform/svg/FontSizes';
import { ScoreRenderer } from '@src/rendering/ScoreRenderer';
import type { Settings } from '@src/Settings';
import { Logger } from '@src/Logger';
import { Environment } from '@src/Environment';
import { SettingsSerializer } from '@src/generated/SettingsSerializer';

/**
 * @target web
 */
export class AlphaTabWebWorker {
    private _renderer!: ScoreRenderer;
    private _main: IWorkerScope;

    public constructor(main: IWorkerScope) {
        this._main = main;
        this._main.addEventListener('message', this.handleMessage.bind(this), false);
    }

    public static init(): void {
        (Environment.globalThis as any).alphaTabWebWorker = new AlphaTabWebWorker(
            Environment.globalThis as IWorkerScope
        );
    }

    private handleMessage(e: MessageEvent): void {
        const data: any = e.data;
        const cmd: any = data ? data.cmd : '';
        switch (cmd) {
            case 'alphaTab.initialize':
                const settings: Settings = JsonConverter.jsObjectToSettings(data.settings);
                Logger.logLevel = settings.core.logLevel;
                this._renderer = new ScoreRenderer(settings);
                this._renderer.partialRenderFinished.on(result => {
                    this._main.postMessage({
                        cmd: 'alphaTab.partialRenderFinished',
                        result: result
                    });
                });
                this._renderer.partialLayoutFinished.on(result => {
                    this._main.postMessage({
                        cmd: 'alphaTab.partialLayoutFinished',
                        result: result
                    });
                });
                this._renderer.renderFinished.on(result => {
                    this._main.postMessage({
                        cmd: 'alphaTab.renderFinished',
                        result: result
                    });
                });
                this._renderer.postRenderFinished.on(() => {
                    this._main.postMessage({
                        cmd: 'alphaTab.postRenderFinished',
                        boundsLookup: this._renderer.boundsLookup?.toJson() ?? null
                    });
                });
                this._renderer.preRender.on(resize => {
                    this._main.postMessage({
                        cmd: 'alphaTab.preRender',
                        resize: resize
                    });
                });
                this._renderer.error.on(this.error.bind(this));
                break;
            case 'alphaTab.invalidate':
                this._renderer.render();
                break;
            case 'alphaTab.resizeRender':
                this._renderer.resizeRender();
                break;
            case 'alphaTab.renderResult':
                this._renderer.renderResult(data.resultId);
                break;
            case 'alphaTab.setWidth':
                this._renderer.width = data.width;
                break;
            case 'alphaTab.renderScore':
                this.updateFontSizes(data.fontSizes);
                const score: any =
                    data.score == null ? null : JsonConverter.jsObjectToScore(data.score, this._renderer.settings);
                this.renderMultiple(score, data.trackIndexes);
                break;
            case 'alphaTab.updateSettings':
                this.updateSettings(data.settings);
                break;
        }
    }

    private updateFontSizes(fontSizes: { [key: string]: FontSizeDefinition } | Map<string, FontSizeDefinition>): void {
        if (!(fontSizes instanceof Map)) {
            const obj = fontSizes;
            fontSizes = new Map<string, FontSizeDefinition>();
            for (const font in obj) {
                fontSizes.set(font, obj[font]);
            }
        }

        if (fontSizes) {
            if (!FontSizes.FontSizeLookupTables) {
                FontSizes.FontSizeLookupTables = new Map<string, FontSizeDefinition>();
            }
            for (const [k, v] of fontSizes) {
                FontSizes.FontSizeLookupTables.set(k, v);
            }
        }
    }

    private updateSettings(json: unknown): void {
        SettingsSerializer.fromJson(this._renderer.settings, json);
    }

    private renderMultiple(score: Score | null, trackIndexes: number[] | null): void {
        try {
            this._renderer.renderScore(score, trackIndexes);
        } catch (e) {
            this.error(e as Error);
        }
    }

    private error(error: Error): void {
        Logger.error('Worker', 'An unexpected error occurred in worker', error);
        this._main.postMessage({
            cmd: 'alphaTab.error',
            error: error
        });
    }
}
